src/Spice86.Core/Emulator/InterruptHandlers/Dos/Ems/ExpandedMemoryManager.cs

```diff
@@ -0,0 +1,1220 @@
+    /// EMM system handle (reserved for OS usage)
+    /// </summary>
+    public const byte EmmSystemHandle = 0;
+
+    public const byte EmmMaxHandles = 16;
+
+    public const uint EmmPageFrameSize = 64 * 1024;
+
+    public const byte EmmMaxPhysicalPages = (byte) (EmmPageFrameSize / EmmPageSize);
+
+    public const ushort EmmNullPage = 0xFFFF;
+    
+    public const ushort EmmNullHandle = 0xFFFF;
+
+    public const ushort EmmPageFrameSegment = 0xE000;
+
+    public const ushort EmmPageSize = 16384;
+
+    public override ushort? InterruptHandlerSegment => 0xF100;
+    
+    public override byte Index => 0x67;
+
+    private readonly ILoggerService _loggerService;
+    
+    public MemoryBlock MemoryBlock { get; }
+
+    public EmmMapping[] EmmSegmentMappings { get; } = new EmmMapping[64];
+
+    public EmmMapping[] EmmMappings { get; } = new EmmMapping[EmmHandle.EmmMaxPhysicalPages];
+    
+    public EmmHandle[] EmmHandles { get; } = new EmmHandle[EmmMaxHandles];
+    
+    public ushort MemorySizeInMb { get; }
+
+    public int TotalPages => MemoryBlock.Pages;
+
+    private readonly Ram _ram;
+
+    public ushort GetFreePages() => Math.Min((ushort) 0x7FFF, (ushort) (GetFreeMemoryTotal() / 4));
+
+    public ExpandedMemoryManager(Machine machine, ILoggerService loggerService) : base(machine, loggerService) {
+        _loggerService = loggerService;
+        MemorySizeInMb = Math.Max((ushort)8, (ushort) (_memory.Size / 1024 / 1024));
+        _ram = new Ram((uint) (MemorySizeInMb * 1024 * 1024));
+        var device = new CharacterDevice(DeviceAttributes.Ioctl, EmsIdentifier);
+        machine.Dos.AddDevice(device, InterruptHandlerSegment, 0x0000);
+        for (int i = 0; i < EmmHandles.Length; i++) {
+            EmmHandles[i] = new EmmHandle();
+        }
+
+        for (int i = 0; i < EmmMappings.Length; i++) {
+            EmmMappings[i] = new EmmMapping();
+        }
+
+        for (int i = 0; i < EmmSegmentMappings.Length; i++) {
+            EmmSegmentMappings[i] = new EmmMapping();
+        }
+
+        MemoryBlock = new MemoryBlock(MemorySizeInMb);
+
+        FillDispatchTable();
+        
+        _state.AH = EmmAllocateSystemHandle(8);
+    }
+    
+    /// <summary>
+    /// Allocates OS-dedicated handle (ems handle zero, 128kb) <br/>
+    /// This handle is never deallocated.
+    /// </summary>
+    /// <param name="pages">The number of pages to initialize</param>
+    /// <returns>Success or error code.</returns>
+    private byte EmmAllocateSystemHandle(ushort pages) {
+        /* Check for enough free pages */
+        if ((GetFreeMemoryTotal() / 4) < pages) {
+            if (_loggerService.IsEnabled(LogEventLevel.Warning))
+                _loggerService.Warning("EMS: Not enough free pages to allocate system handle. Free pages: {0}, requested pages: {1}", GetFreeMemoryTotal() / 4, pages);
+            return EmmStatus.EmmOutOfLogicalPages;
+        }
+
+        /* Release memory if already allocated */
```

Comment:
How can this happen? This is only called in constructor right?

Reply:
It depends on how much main memory we have. Below 1 MB this won't work.

Reply:
In dosbox ems can be disabled or enabled at any time. This is probably why.